NTM Optimization Update!

BallOfEnergy1 opened the pull request:
The NTM Optimization Update is here!

Changes:
Completely removed the `NBTPacket` and `INBTPacketReceiver` classes in favor of using `ByteBuf`s for everything.
Reworked RBMK and Chicago Pile neutron calculations, now they should be way more performant (and use the same system).
Tons of other small tweaks.

Profiler results:
100 Rod RBMK (made by nootles):
[Before](https://spark.lucko.me/rAKONjbyqO): 22.9 avg MSPT
[After](https://spark.lucko.me/5de9F91p7V): 2.58 avg MSPT
Gerver (my personal server):
[Before](https://spark.lucko.me/VKOljXLoNW): 12.9 avg MSPT
[After](https://spark.lucko.me/GSFlp41LVz): 5.32 avg MSPT

Don't merge yet, still looking for a good place to check for every possible issue that could exist.
(sorry about the crazy commit structure, GitHub was being an asshole and wouldn't let me resolve conflicts)

BallOfEnergy1 committed:
holy shiiiiiiiiiitttt RBMK optimization update

BallOfEnergy1 committed:
holy shiiiiiiiiiitttt RBMK optimization update

BallOfEnergy1 committed:
fuck!

BallOfEnergy1 committed:
fuck!

BallOfEnergy1 committed:
this is fucking insane dude
RBMK optimization update soon!!!!!!!

BallOfEnergy1 committed:
this is fucking insane dude
RBMK optimization update soon!!!!!!!

BallOfEnergy1 committed:
hooolyyyy shiiiiiiiit

i finally fixed the bug!!!!!!!!!!

BallOfEnergy1 committed:
hooolyyyy shiiiiiiiit

i finally fixed the bug!!!!!!!!!!

BallOfEnergy1 committed:
99.999% there
just some more optimization!
(node culling)

BallOfEnergy1 committed:
99.999% there
just some more optimization!
(node culling)

BallOfEnergy1 committed:
oops

BallOfEnergy1 committed:
oops

BallOfEnergy1 committed:
i am going to kms

BallOfEnergy1 committed:
i am going to kms

BallOfEnergy1 committed:
i fixed it i fixed it i fixed it i fixed it i fixed it i fixed it
fucking FINALLY

BallOfEnergy1 committed:
i fixed it i fixed it i fixed it i fixed it i fixed it i fixed it
fucking FINALLY

BallOfEnergy1 committed:
Merge remote-tracking branch 'upstream/master' into RBMK_optimization

# Conflicts:
#	src/main/java/com/hbm/tileentity/machine/rbmk/TileEntityRBMKRod.java

BallOfEnergy1 committed:
fix up some imports!

BallOfEnergy1 committed:
fix up some imports!

HbmMods committed:
Merge master into branch Optimization, intelliJ couldn't do it, so I did it myself.

HbmMods committed:
Merge master into branch Optimization, intelliJ couldn't do it, so I did it myself.

BallOfEnergy1 committed:
more shit

BallOfEnergy1 committed:
more shit

BallOfEnergy1 committed:
i want to kill myself :heart:

BallOfEnergy1 committed:
i want to kill myself :heart:

BallOfEnergy1 committed:
oh my fucking god

BallOfEnergy1 committed:
oh my fucking god

BallOfEnergy1 committed:
the ByteBuf shenanigans begin!!

BallOfEnergy1 committed:
the ByteBuf shenanigans begin!!

BallOfEnergy1 committed:
holy shit what the fuck am i smoking

BallOfEnergy1 committed:
holy shit what the fuck am i smoking

BallOfEnergy1 committed:
part 1

BallOfEnergy1 committed:
part 1

BallOfEnergy1 committed:
holy shit im done (part 2)

BallOfEnergy1 committed:
holy shit im done (part 2)

BallOfEnergy1 committed:
Merge remote-tracking branch 'origin/Optimization' into Optimization

# Conflicts:
#	src/main/java/com/hbm/tileentity/TileEntityMachineBase.java
#	src/main/java/com/hbm/tileentity/TileEntityTickingBase.java
#	src/main/java/com/hbm/tileentity/machine/TileEntityMachineAutosaw.java
#	src/main/java/com/hbm/tileentity/machine/TileEntityMachineRadarScreen.java
#	src/main/java/com/hbm/tileentity/machine/rbmk/TileEntityRBMKRod.java
#	src/main/java/com/hbm/tileentity/network/TileEntityRequestNetworkContainer.java

BallOfEnergy1 committed:
fixes after the merge conflicts (death)

BallOfEnergy1 committed:
Unused import cleaning!

BallOfEnergy1 committed:
Merge branch 'HbmMods:master' into Optimization

BallOfEnergy1 committed:
Merge branch 'HbmMods:master' into Optimization

BallOfEnergy1 committed:
Merge branch 'HbmMods:master' into Optimization

BallOfEnergy1 committed:
Merge remote-tracking branch 'upstream/master' into Optimization

# Conflicts:
#	src/main/java/com/hbm/tileentity/bomb/TileEntityNukeBalefire.java
#	src/main/java/com/hbm/tileentity/machine/TileEntityAshpit.java
#	src/main/java/com/hbm/tileentity/machine/TileEntityChungus.java
#	src/main/java/com/hbm/tileentity/machine/TileEntityCondenser.java
#	src/main/java/com/hbm/tileentity/machine/TileEntityCore.java
#	src/main/java/com/hbm/tileentity/machine/TileEntityDiFurnace.java
#	src/main/java/com/hbm/tileentity/machine/TileEntityDiFurnaceRTG.java
#	src/main/java/com/hbm/tileentity/machine/TileEntityElectrolyser.java
#	src/main/java/com/hbm/tileentity/machine/TileEntityFurnaceIron.java
#	src/main/java/com/hbm/tileentity/machine/TileEntityFurnaceSteel.java
#	src/main/java/com/hbm/tileentity/machine/TileEntityHeatBoiler.java
#	src/main/java/com/hbm/tileentity/machine/TileEntityHeatBoilerIndustrial.java
#	src/main/java/com/hbm/tileentity/machine/TileEntityHeaterHeatex.java
#	src/main/java/com/hbm/tileentity/machine/TileEntityMachineAutosaw.java
#	src/main/java/com/hbm/tileentity/machine/TileEntityMachineCombustionEngine.java
#	src/main/java/com/hbm/tileentity/machine/TileEntityMachineCompressor.java
#	src/main/java/com/hbm/tileentity/machine/TileEntityMachineDiesel.java
#	src/main/java/com/hbm/tileentity/machine/TileEntityMachineEPress.java
#	src/main/java/com/hbm/tileentity/machine/TileEntityMachineHephaestus.java
#	src/main/java/com/hbm/tileentity/machine/TileEntityMachineMixer.java
#	src/main/java/com/hbm/tileentity/machine/TileEntityMachinePlasmaHeater.java
#	src/main/java/com/hbm/tileentity/machine/TileEntityMachinePress.java
#	src/main/java/com/hbm/tileentity/machine/TileEntityMachinePumpBase.java
#	src/main/java/com/hbm/tileentity/machine/TileEntityMachineRadGen.java
#	src/main/java/com/hbm/tileentity/machine/TileEntityMachineRadiolysis.java
#	src/main/java/com/hbm/tileentity/machine/TileEntityMachineSchrabidiumTransmutator.java
#	src/main/java/com/hbm/tileentity/machine/TileEntityMachineSolderingStation.java
#	src/main/java/com/hbm/tileentity/machine/TileEntityMachineStrandCaster.java
#	src/main/java/com/hbm/tileentity/machine/TileEntityMachineTurbofan.java
#	src/main/java/com/hbm/tileentity/machine/TileEntitySoyuzLauncher.java
#	src/main/java/com/hbm/tileentity/machine/TileEntitySteamEngine.java
#	src/main/java/com/hbm/tileentity/machine/oil/TileEntityMachineCatalyticCracker.java
#	src/main/java/com/hbm/tileentity/machine/oil/TileEntityMachineCatalyticReformer.java
#	src/main/java/com/hbm/tileentity/machine/oil/TileEntityMachineCoker.java
#	src/main/java/com/hbm/tileentity/machine/oil/TileEntityMachineGasFlare.java
#	src/main/java/com/hbm/tileentity/machine/oil/TileEntityMachineLiquefactor.java
#	src/main/java/com/hbm/tileentity/machine/oil/TileEntityMachineRefinery.java
#	src/main/java/com/hbm/tileentity/machine/oil/TileEntityMachineSolidifier.java
#	src/main/java/com/hbm/tileentity/machine/oil/TileEntityMachineVacuumDistill.java
#	src/main/java/com/hbm/tileentity/machine/rbmk/TileEntityRBMKConsole.java
#	src/main/java/com/hbm/tileentity/machine/rbmk/TileEntityRBMKControlAuto.java
#	src/main/java/com/hbm/tileentity/machine/storage/TileEntityMassStorage.java
#	src/main/java/com/hbm/tileentity/network/TileEntityCraneRouter.java
#	src/main/java/com/hbm/tileentity/network/TileEntityDroneCrate.java
#	src/main/java/com/hbm/tileentity/network/TileEntityDroneRequester.java
#	src/main/java/com/hbm/tileentity/turret/TileEntityTurretFritz.java
#	src/main/java/com/hbm/tileentity/turret/TileEntityTurretMaxwell.java
#	src/main/java/com/hbm/tileentity/turret/TileEntityTurretRichard.java
#	src/main/java/com/hbm/tileentity/turret/TileEntityTurretSentry.java
#	src/main/java/com/hbm/tileentity/turret/TileEntityTurretTauon.java

BallOfEnergy1 committed:
Merge remote-tracking branch 'upstream/master' into Optimization

# Conflicts:
#	src/main/java/com/hbm/tileentity/TileEntityMachineBase.java
#	src/main/java/com/hbm/tileentity/machine/TileEntityHeaterElectric.java
#	src/main/java/com/hbm/tileentity/machine/TileEntityMachineSolderingStation.java
#	src/main/java/com/hbm/tileentity/machine/TileEntityPWRController.java
#	src/main/java/com/hbm/tileentity/machine/rbmk/TileEntityCraneConsole.java
#	src/main/java/com/hbm/tileentity/network/TileEntityCraneGrabber.java

BallOfEnergy1 committed:
yeah these are commits from like a month ago idk what any of this is

BallOfEnergy1 committed:
Merge remote-tracking branch 'upstream/master' into Optimization

BallOfEnergy1 committed:
Holy SHIT! i think i fixed it

BallOfEnergy1 committed:
there i fixed it :woo:

BallOfEnergy1 committed:
bug fixes part 1
(thank you radium)

BallOfEnergy1 committed:
bug fixes part 2
(both me and radium did testing)

BallOfEnergy1 committed:
Small fixes

BallOfEnergy1 committed:
Merge remote-tracking branch 'upstream/master' into Optimization

BallOfEnergy1 committed:
just some final tweaks to make sure that I don't fuck anything up

MellowArpeggiation commented:
![cb77ea5a2d1587c8afc1c43a8352349c](https://github.com/user-attachments/assets/a672a73c-ba1f-48fb-805b-c3c13e20f9ff)
nyoom

BallOfEnergy1 commented:
ball

legendarydoge30 commented:
chat is this real

BallOfEnergy1 committed:
Upgrade rework, now with 3463% more caching!
(it was slow and needed a redo)

BallOfEnergy1 committed:
Merge remote-tracking branch 'upstream/master' into Optimization

# Conflicts:
#	src/main/java/com/hbm/tileentity/network/TileEntityCraneGrabber.java

HbmMods commented:
Do you have any profiler data on the upgrade stuff? Because I'm not sure if caching that would even change a whole lot, considering often times machines run without upgrades anyway and machines only have 2-4 upgrade slots at most.

BallOfEnergy1 committed:
Merge branch 'HbmMods:master' into Optimization

RedstoneParkour commented:
would it be possible to process rbmks that can't interact with each other in seperate threads for more speed? or did you already implement that?

BallOfEnergy1 commented:
Multi threading RBMKs is possible, though it would have such a massive overhead for synchronizing them that it would be slower than with what we have right now.

BallOfEnergy1 commented:
> Do you have any profiler data on the upgrade stuff? Because I'm not sure if caching that would even change a whole lot, considering often times machines run without upgrades anyway and machines only have 2-4 upgrade slots at most.

The main reason I even touched upgrades at all was because 70000hp mentioned that he noticed some speed issues when using that system since it checks every tick every upgrade slot, though I haven't seen any performance impacts of my own.

HbmMods commented:
Upgrade checks are just for loops the size of the upgrade slot range (single digit) which either NOPs when no upgrades are installed or adding upgrade info to a hashmap (with at most as many elements as there are occupied slots), whatever the machine itself does (recipe lookup, port connections, etc) is likely magnitudes more impactful than that, which is why I'd prefer seeing profiler results first before changing anything with the upgrade system.

BallOfEnergy1 commented:
Alright, I'll try and see if I can get some profiler data on it this afternoon, as well as maybe looking into how to optimize other parts of machines.

BallOfEnergy1 committed:
Merge remote-tracking branch 'upstream/master' into Optimization

# Conflicts:
#	src/main/java/com/hbm/main/ModEventHandlerRenderer.java

BallOfEnergy1 committed:
Merge remote-tracking branch 'origin/Optimization' into Optimization

BallOfEnergy1 committed:
pushing to prod without testing
woo!!

BallOfEnergy1 committed:
Merge remote-tracking branch 'upstream/master' into Optimization

# Conflicts:
#	src/main/java/com/hbm/main/ModEventHandlerRenderer.java
#	src/main/java/com/hbm/tileentity/machine/TileEntityPWRController.java

BallOfEnergy1 committed:
Make ExtPropPacket ByteBuf based, along with some other performance improvements such as removing debug fluid net code and making RBMKDials more accepting of new gamerules.

BallOfEnergy1 committed:
Merge branch 'HbmMods:master' into Optimization

BallOfEnergy1 committed:
Merge remote-tracking branch 'upstream/master' into Optimization

# Conflicts:
#	src/main/java/com/hbm/extprop/HbmLivingProps.java
#	src/main/java/com/hbm/handler/EntityEffectHandler.java

BallOfEnergy1 committed:
Merge branch 'HbmMods:master' into Optimization

BallOfEnergy1 committed:
there's actually so many optimizations here that it'd be impossible for be to explain them all

BallOfEnergy1 commented:
With this most recent commit, the 100 rod RBMK time has been reduced to a mere [0.818 MSPT avg](https://spark.lucko.me/nenpgi9qdQ), as well as the memory usage being reduced by almost double due to optimizations with how the `BlockPos` instances are handled. This has not been tested fully with the gerver save yet, however my tests from a while ago suggest a 2x improvement over the *already optimized* version.

BallOfEnergy1 committed:
Merge branch 'HbmMods:master' into Optimization

BallOfEnergy1 committed:
Merge branch 'HbmMods:master' into Optimization

BallOfEnergy1 committed:
Merge branch 'HbmMods:master' into Optimization

BallOfEnergy1 committed:
More shit, also precompiled packets.
also holy shit the whitespace changes, intelliJ wanted to optimize the imports and refactored a ton of whitespace in the process.

BallOfEnergy1 committed:
Merge remote-tracking branch 'origin/Optimization' into Optimization

BallOfEnergy1 committed:
Merge remote-tracking branch 'upstream/master' into Optimization

# Conflicts:
#	src/main/java/com/hbm/main/ResourceManager.java

BallOfEnergy1 commented:
> With this most recent commit, the 100 rod RBMK time has been reduced to a mere [0.818 MSPT avg](https://spark.lucko.me/nenpgi9qdQ), as well as the memory usage being reduced by almost double due to optimizations with how the `BlockPos` instances are handled. This has not been tested fully with the gerver save yet, however my tests from a while ago suggest a 2x improvement over the _already optimized_ version.

This has since been improved to ~0.616 MSPT med, as well as memory usage being reduced from ~800MB before any optimizations and ~2100MB after MSPT optimizations, down to ~300MB (All tested on the 100 rod RBMK save).

BallOfEnergy1 committed:
oops, bugfixes!
- Pooled buffers removed due to issues with releasing the memory.
- ExtProps fixed for players/entities; improper deserialization.

BallOfEnergy1 committed:
small flixes

BallOfEnergy1 committed:
Merge remote-tracking branch 'refs/remotes/upstream/master' into Optimization

# Conflicts:
#	src/main/java/com/hbm/main/ModEventHandlerClient.java
#	src/main/java/com/hbm/main/ResourceManager.java
#	src/main/java/com/hbm/util/EntityDamageUtil.java

BallOfEnergy1 committed:
uh oh

BallOfEnergy1 committed:
Merge remote-tracking branch 'refs/remotes/upstream/master' into Optimization

# Conflicts:
#	src/main/java/com/hbm/util/CompatExternal.java
#	src/main/java/com/hbm/util/CrucibleUtil.java
#	src/main/java/com/hbm/util/EntityDamageUtil.java

BallOfEnergy1 committed:
Merge remote-tracking branch 'main/master' into Optimization

# Conflicts:
#	src/main/java/com/hbm/blocks/ModBlocks.java
#	src/main/java/com/hbm/handler/EntityEffectHandler.java
#	src/main/java/com/hbm/main/ClientProxy.java
#	src/main/java/com/hbm/main/CraftingManager.java
#	src/main/java/com/hbm/main/MainRegistry.java
#	src/main/java/com/hbm/main/ModEventHandlerClient.java
#	src/main/java/com/hbm/main/ModEventHandlerRenderer.java
#	src/main/java/com/hbm/main/ResourceManager.java
#	src/main/java/com/hbm/tileentity/turret/TileEntityTurretBaseNT.java
#	src/main/java/com/hbm/tileentity/turret/TileEntityTurretRichard.java
#	src/main/java/com/hbm/tileentity/turret/TileEntityTurretSentry.java
#	src/main/java/com/hbm/tileentity/turret/TileEntityTurretTauon.java
#	src/main/java/com/hbm/util/EntityDamageUtil.java

HbmMods commented:
can you take all the file changes out that just change whitespace and import order? bit hard to review the PR when it's 90% air

BallOfEnergy1 committed:
Merge remote-tracking branch 'upstream/master' into Optimization

# Conflicts:
#	src/main/java/com/hbm/blocks/ModBlocks.java
#	src/main/java/com/hbm/handler/EntityEffectHandler.java
#	src/main/java/com/hbm/main/ClientProxy.java
#	src/main/java/com/hbm/main/CraftingManager.java
#	src/main/java/com/hbm/main/MainRegistry.java
#	src/main/java/com/hbm/main/ModEventHandlerClient.java
#	src/main/java/com/hbm/main/ModEventHandlerRenderer.java
#	src/main/java/com/hbm/main/NEIConfig.java
#	src/main/java/com/hbm/main/ResourceManager.java
#	src/main/java/com/hbm/tileentity/turret/TileEntityTurretBaseNT.java
#	src/main/java/com/hbm/tileentity/turret/TileEntityTurretRichard.java
#	src/main/java/com/hbm/tileentity/turret/TileEntityTurretSentry.java
#	src/main/java/com/hbm/tileentity/turret/TileEntityTurretTauon.java
#	src/main/java/com/hbm/util/CompatExternal.java
#	src/main/java/com/hbm/util/CrucibleUtil.java
#	src/main/java/com/hbm/util/EntityDamageUtil.java

BallOfEnergy1 committed:
Merge remote-tracking branch 'origin/Optimization' into Optimization

# Conflicts:
#	src/main/java/com/hbm/main/ModEventHandlerClient.java
#	src/main/java/com/hbm/util/CompatExternal.java

BallOfEnergy1 commented:
> can you take all the file changes out that just change whitespace and import order? bit hard to review the PR when it's 90% air

I genuinely have no easy way of doing this other than reviewing every file by hand and manually fixing whitespace, it all comes down to the `.editorconfig` martin added and intellIJ taking that as rule (then reformatting the entire src with it). This can be continued in discord as it's easier to communicate there.

Ashen-sysadmin commented:
![image](https://github.com/user-attachments/assets/883a513d-33b4-4d1a-9a47-2a0e4f27e98e)
Uh. this button exists. The white-space changes are unavoidable because of the autoconfig system set up for intelliJ

HbmMods commented:
ok correction the whitespace thing isn't that bad, it's the imports changing. no biggie i'll just do those files first then

HbmMods committed:
Merge branch 'master' into Optimization

BallOfEnergy1 committed:
Merge branch 'HbmMods:master' into Optimization

BallOfEnergy1 committed:
More RAM optimizations with buffer releasing. Also fix a bug in RBMK processing and a few merge-related errors.

BallOfEnergy1 committed:
Merge remote-tracking branch 'origin/Optimization' into Optimization

BallOfEnergy1 committed:
Merge branch 'HbmMods:master' into Optimization

BallOfEnergy1 committed:
Packet threading debug command!!
(Also packet threading config option)

BallOfEnergy1 committed:
an absolutely goofy command I'd say (also yet another failsafe, this time for a ton of packet discarding.)

BallOfEnergy1 committed:
fix some stuff, also fix sending exactly 19x more packets than needed lmao

BallOfEnergy1 committed:
well crap this was an oops

BallOfEnergy1 committed:
fucky wucky (kms)

BallOfEnergy1 committed:
Merge remote-tracking branch 'upstream/master' into Optimization

# Conflicts:
#	src/main/java/com/hbm/blocks/ModBlocks.java
#	src/main/java/com/hbm/lib/HbmCollection.java
#	src/main/java/com/hbm/main/ClientProxy.java
#	src/main/java/com/hbm/main/MainRegistry.java
#	src/main/java/com/hbm/main/ModEventHandler.java
#	src/main/java/com/hbm/main/NEIConfig.java
#	src/main/java/com/hbm/util/LootGenerator.java

BallOfEnergy1 committed:
Merge remote-tracking branch 'upstream/master' into Optimization

# Conflicts:
#	src/main/java/com/hbm/main/ResourceManager.java

BallOfEnergy1 committed:
fucky wucky pt. 2

BallOfEnergy1 committed:
fucky wucky pt. 3

BallOfEnergy1 committed:
am i retarded
fucky wucky pt. 4

BallOfEnergy1 committed:
Merge branch 'HbmMods:master' into Optimization

BallOfEnergy1 committed:
Chicago Pile fix ClassCastException.

BallOfEnergy1 committed:
Fix /ntmpackets command registration.
Fix small race condition with a ReentrantLock.

BallOfEnergy1 committed:
An absolute ton of cleanup.

HbmMods committed:
Merge branch 'master' into Optimization

HbmMods commented on src/main/java/com/hbm/tileentity/machine/TileEntityBroadcaster.java:
```diff
@@ -1,55 +1,103 @@
+
+			int intendedVolume = 25;
+
+			EntityClientPlayerMP player = Minecraft.getMinecraft().thePlayer;
```
please use `MainRegistry.proxy.me()` for this, `Minecraft.getMinecraft()` and any other client-only calls like that are a big nono even in `world.isRemote` blocks

BallOfEnergy1 committed:
Fix TileEntityBroadcaster.

BallOfEnergy1 committed:
Merge remote-tracking branch 'origin/Optimization' into Optimization

BallOfEnergy1 committed:
Wow, that was bad.
Fix fluids randomly disappearing/appearing in the Heat Exchanging Heater in rare circumstances due to packet compilation times.

BallOfEnergy1 committed:
Small fix for upgrades on TileEntityElectrolyser.

HbmMods commented:
let's goo

Xenox003 commented:
@BallOfEnergy1 What version of IntelliJ did you use here?
I could not get this project set up in my IntellJ-ENV cause of the old gradle version.
So i'm just curious.

BallOfEnergy1 commented:
Like some old old version from 2021... You only really need to get the gradle version right though.
